### PR TITLE
Added a new AST Walker that instruments the AST to provide callbacks that simulate a program counter

### PR DESCRIPTION
**Pull request by mxswd**
Based off the PlaygroundTransform, this new ASTWalker leaves calls to `__builtin_pc_before` and `__builtin_pc_after` before and after a user would expect a program counter to enter a range of source code.

### CODE REVIEW

**Comment by zisko** [comment]
@swift-ci please test

File: include/swift/Subsystems.h
Comment (jrose-apple) [inline include/swift/Subsystems.h]: This still says the playground transform would be complete, but you now run it before.

File: lib/Sema/InstrumenterSupport.cpp
Comment (jrose-apple) [inline lib/Sema/InstrumenterSupport.cpp]: Nitpick: instead of putting the destructor out of line, add a dummy "anchor" method. This way destruction can still be inlined at the use site.

(Also, prefer `= default` to `{}` for destructors that don't do anything.)

File: lib/Sema/InstrumenterSupport.cpp
Comment (jrose-apple) [inline lib/Sema/InstrumenterSupport.cpp]: Nitpick: We don't generally reopen namespaces within .cpp files. Instead, just add another `using namespace`, or fully qualify the method name.

File: lib/Sema/InstrumenterSupport.h
Comment (jrose-apple) [inline lib/Sema/InstrumenterSupport.h]: Nitpick: please add the `-*- c++ -*-` header here, for the ~poor souls~ nice folks using emacs.

File: include/swift/Frontend/FrontendOptions.h
Comment (slavapestov) [inline include/swift/Frontend/FrontendOptions.h]: This needs a better comment

File: lib/Sema/PCMacro.cpp
Comment (jrose-apple) [inline lib/Sema/PCMacro.cpp]: This needs more characters; might as well fit the whole `unsigned`.

File: lib/Sema/PCMacro.cpp
Comment (jrose-apple) [inline lib/Sema/PCMacro.cpp]: If you want to look through sugar, use `InitExpr->getType()->getAs<LValueType>` rather than this. (Most of the time that's what you want.)

**Review by jrose-apple — COMMENTED** [review]
I'll let @scallanan do the main review, though I threw in a few comments. The part hooking it up to the rest of the compiler looks good to me.

File: include/swift/Frontend/FrontendOptions.h
Comment (mxswd) [inline include/swift/Frontend/FrontendOptions.h]: Sure thing! Updated.

File: include/swift/Subsystems.h
Comment (mxswd) [inline include/swift/Subsystems.h]: Hah yeah! Forgot I changed that.

File: lib/Sema/InstrumenterSupport.cpp
Comment (mxswd) [inline lib/Sema/InstrumenterSupport.cpp]: Done!

File: lib/Sema/InstrumenterSupport.cpp
Comment (mxswd) [inline lib/Sema/InstrumenterSupport.cpp]: The more you know

File: lib/Sema/PCMacro.cpp
Comment (mxswd) [inline lib/Sema/PCMacro.cpp]: Oops. Fixed in Playground Transform as well.

Good job JordASAN!

File: lib/Sema/PCMacro.cpp
Comment (mxswd) [inline lib/Sema/PCMacro.cpp]: Cool! Updated in PlaygroundTransform as well.

**Comment by mxswd** [comment]
Updated with feedback!

File: lib/Sema/InstrumenterSupport.h
Comment (scallanan) [inline lib/Sema/InstrumenterSupport.h]: Do you need this using declaration if everything in the header is inside that namespace?  You're opting everyone who includes it into swift.

File: lib/Sema/InstrumenterSupport.h
Comment (scallanan) [inline lib/Sema/InstrumenterSupport.h]: The fact that there's all this code in a header file is a smell.  Can we have a very simple function that calls through to an Impl and then does the dyn_cast to Added<T> as a service?

File: lib/Sema/PlaygroundTransform.cpp
Comment (scallanan) [inline lib/Sema/PlaygroundTransform.cpp]: Did clang-format really let that double-space before TmpNameIndex through??

File: lib/Sema/InstrumenterSupport.h
Comment (mxswd) [inline lib/Sema/InstrumenterSupport.h]: I can't seem to figure out how to do this. If the Impl's type is:
```
bool InstrumenterBase::doTypeCheckImpl(ASTContext &Ctx, DeclContext *DC,
                                       Added<Expr *> &parsedExpr)
```
Then the code would be:
```
  Expr *E = *parsedExpr;
  TC.typeCheckExpression(E, DC);
  parsedExpr = Added<Expr *>(E);

  if (*parsedExpr) {
    ErrorFinder errorFinder;
    parsedExpr->walk(errorFinder);
```
Wouldn't this fail because, without the T, the `walk` call here call `Expr *`'s `walk` instead of `T`'s `walk`?

File: lib/Sema/PlaygroundTransform.cpp
Comment (mxswd) [inline lib/Sema/PlaygroundTransform.cpp]: No, fixed. :P

File: lib/Sema/InstrumenterSupport.h
Comment (mxswd) [inline lib/Sema/InstrumenterSupport.h]: Oops! Fixed.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
Tests passed, but waiting for @scallanan to give a final approval.

File: lib/Sema/InstrumenterSupport.h
Comment (scallanan) [inline lib/Sema/InstrumenterSupport.h]: Do you mean we're type-checking things that aren't `Expr`s?  I don't think `walk` is implemented by the subclasses – it's a function on `Expr` and that's it.

**Comment by mxswd** [comment]
@scallanan OK corrected it!

File: lib/Sema/InstrumenterSupport.h
Comment (scallanan) [inline lib/Sema/InstrumenterSupport.h]: Can `ErrorGatherer` and `ErrorFinder` move into the .cpp file now?

**Comment by scallanan** [comment]
@swift-ci please test

**Review by scallanan — APPROVED** [review]
Thank you for your work on this, Maxwell.
I approve the changes.

**Pull request merged**
